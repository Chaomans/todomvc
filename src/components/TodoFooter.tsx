import { Filter } from "./models/models";

type TodoFooterProps = {
    remainingTodos: number;
    completedTodos: number;
    activeFilter: string;
    filters: Filter[];
    onClearCompleted: () => void;
    onChangeActiveFilter: (filterName: string) => void;
};

const TodoFooter = ({ remainingTodos, completedTodos, activeFilter, filters, onClearCompleted, onChangeActiveFilter }: TodoFooterProps) => {

    const beforeChangeActiveFilter = (filter: string) =>{
        onChangeActiveFilter(filter);
    }

    return (
        <footer className="footer">
            <span className="todo-count">
                <strong>{remainingTodos}</strong>{" "}
                {remainingTodos === 0 || remainingTodos > 1 ? "items" : "item"} left
            </span>
            {filters.length > 0 && (
                <ul className="filters">
                    {filters.map((filter) =>(
                        <li key={ filter.value }>
                            <a 
                                href="/" 
                                onClick={(e) => {
                                    e.preventDefault();
                                    beforeChangeActiveFilter(filter.value)
                                }} 
                                className={activeFilter === filter.value ? "selected": ""}
                            >
                                {filter.name}
                            </a>
                        </li>
                    ))}
                </ul>
            )}
            {completedTodos > 0 && (
                <button className="clear-completed" onClick={() => onClearCompleted()}>
                    Clear completed
                </button>
            )}
        </footer>
    );
};

export default TodoFooter;
