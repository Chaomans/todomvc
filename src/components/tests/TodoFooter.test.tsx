import "@testing-library/jest-dom";
import { act, render, screen } from "@testing-library/react";
import TodoFooter from "../TodoFooter";

describe("TodoFooter component", () => {
    it("Should display properly", () => {
        const { asFragment } = render(
            <TodoFooter nbTodo={1} onClearCompleted={() => {}} />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("Should display number of items", async () => {
        render(<TodoFooter nbTodo={25} onClearCompleted={() => {}} />);

        expect(await screen.findByText("25")).toBeInTheDocument();
    });

    it.each`
        nbOfTodos | pluralized      | expected
        ${1}      | ${"item left"}  | ${"1 item left"}
        ${0}      | ${"items left"} | ${"0 items left"}
        ${2}      | ${"items left"} | ${"2 items left"}
    `(
        "Should display '$expected' with $nbOfTodos todos",
        async ({ nbOfTodos, pluralized, expected }) => {
            render(
                <TodoFooter nbTodo={nbOfTodos} onClearCompleted={() => {}} />
            );

            expect(
                await screen.findByText(nbItemLeft(nbOfTodos, pluralized))
            ).toBeInTheDocument();
        }
    );

    const nbItemLeft = (nbOfTodos: number, pluralized: string) => {
        return (content: string, element: Element | null) => {
            return (
                +content === nbOfTodos &&
                (element?.parentNode?.textContent?.endsWith(pluralized) ??
                    false)
            );
        };
    };
});
