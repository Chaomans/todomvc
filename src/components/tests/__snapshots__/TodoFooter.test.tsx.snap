// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoFooter component Should display properly 1`] = `
<DocumentFragment>
  <footer
    class="footer"
  >
    <span
      class="todo-count"
    >
      <strong>
        1
      </strong>
       item left
    </span>
    <button
      class="clear-completed"
    >
      Clear completed
    </button>
  </footer>
</DocumentFragment>
`;
